From 8d244f22068ed8dc8b2a262423634e90165a1049 Mon Sep 17 00:00:00 2001
From: Thu Nguyen <thu@os.amperecomputing.com>
Date: Sun, 20 Jul 2025 10:39:06 +0000
Subject: [PATCH] user_channel: Support setUserIsBootStrapState API

To identify the one user account with object path
`/xyz/openbmc_project/user/<username>` which is host by
`xyz.openbmc_project.User.Manager` is bootStrap account.
The property `IsBootStrap` is added to
`xyz.openbmc_project.User.Attributes` D-Bus interface as the design [1].
Support the API to allow set `IsBootStrap` property in
`xyz.openbmc_project.User.Attributes` D-Bus interface.

[1] https://gerrit.openbmc.org/c/openbmc/docs/+/79327

Upstream-Status: Pending

Change-Id: I04eefc607f993d62901a09436c759af505f5aa85
Signed-off-by: Thu Nguyen <thu@os.amperecomputing.com>
---
 user_channel/user_layer.cpp | 12 +++++++
 user_channel/user_layer.hpp | 10 ++++++
 user_channel/user_mgmt.cpp  | 62 +++++++++++++++++++++++++++++++++++++
 user_channel/user_mgmt.hpp  | 11 +++++++
 4 files changed, 95 insertions(+)

diff --git a/user_channel/user_layer.cpp b/user_channel/user_layer.cpp
index 38eced1..162b58d 100644
--- a/user_channel/user_layer.cpp
+++ b/user_channel/user_layer.cpp
@@ -257,4 +257,16 @@ Cc ipmiUserRemoveUserFromNoneIpmiGroupUsers(const std::string& userName)
     return getUserAccessObject().removeUserFromNoneIpmiGroupUsers(userName);
 }
 
+Cc ipmiUserSetUserBootStrapAccountState(const uint8_t userId,
+                                        const bool& bootStrapState)
+{
+    if (!UserAccess::isValidUserId(userId))
+    {
+        return ccParmOutOfRange;
+    }
+
+    return getUserAccessObject().setUserBootStrapAccountState(
+        userId, bootStrapState);
+}
+
 } // namespace ipmi
diff --git a/user_channel/user_layer.hpp b/user_channel/user_layer.hpp
index 6d8e828..75b5b30 100644
--- a/user_channel/user_layer.hpp
+++ b/user_channel/user_layer.hpp
@@ -296,4 +296,14 @@ Cc ipmiUserAddUserToNonIpmiGroupUsers(const std::string& userName);
  */
 Cc ipmiUserRemoveUserFromNoneIpmiGroupUsers(const std::string& userName);
 
+/** @brief sets wherether user is BootStrap account
+ *
+ *  @param[in] userId - user id
+ *  @param[in] bootStrapState - is boot strap account state
+ *
+ *  @return ccSuccess for success, others for failure.
+ */
+Cc ipmiUserSetUserBootStrapAccountState(const uint8_t userId,
+                                        const bool& bootStrapState);
+
 } // namespace ipmi
diff --git a/user_channel/user_mgmt.cpp b/user_channel/user_mgmt.cpp
index 40c31cc..6a3c515 100644
--- a/user_channel/user_mgmt.cpp
+++ b/user_channel/user_mgmt.cpp
@@ -1834,4 +1834,66 @@ Cc UserAccess::removeUserFromNoneIpmiGroupUsers(const std::string& userName)
     return ccSuccess;
 }
 
+Cc UserAccess::setUserBootStrapAccountState(const uint8_t userId,
+                                            const bool& bootStrapState)
+{
+    if (!isValidUserId(userId))
+    {
+        return ccParmOutOfRange;
+    }
+
+    std::string userName;
+    if (ipmiUserGetUserName(userId, userName) != ccSuccess)
+    {
+        lg2::error("User Name not found, user Id: {USER_ID}", "USER_ID",
+                   userId);
+        return ccParmOutOfRange;
+    }
+    if (userName.empty())
+    {
+        lg2::error("User name is empty.");
+        return ccUnspecifiedError;
+    }
+
+    sdbusplus::message::object_path tempUserPath(userObjBasePath);
+    tempUserPath /= userName;
+    std::string userPath(tempUserPath);
+
+    try
+    {
+        setDbusProperty(bus, userMgrService, userPath, usersInterface,
+                        userIsBootStrapProperty, bootStrapState);
+    }
+    catch (const std::exception& e)
+    {
+        lg2::error(
+            "Can't set value of {PROP} property in the {INF} interface at path {PATH} error {ERROR}",
+            "PROP", userIsBootStrapProperty, "INF", usersInterface, "PATH",
+            userPath, "ERROR", e);
+    }
+
+    boost::interprocess::scoped_lock<boost::interprocess::named_recursive_mutex>
+        userLock{*userMutex};
+    UserInfo* userInfo = getUserInfo(userId);
+    if (!userInfo)
+    {
+        lg2::error("Failed to get user info of user name {NAME}", "NAME",
+                   userName);
+        return ccUnspecifiedError;
+    }
+    userInfo->fixedUserName = true;
+
+    try
+    {
+        writeUserData();
+    }
+    catch (const std::exception& e)
+    {
+        lg2::error("Write user data failed");
+        return ccUnspecifiedError;
+    }
+
+    return ccSuccess;
+}
+
 } // namespace ipmi
diff --git a/user_channel/user_mgmt.hpp b/user_channel/user_mgmt.hpp
index 2a74f9f..fc62960 100644
--- a/user_channel/user_mgmt.hpp
+++ b/user_channel/user_mgmt.hpp
@@ -76,6 +76,7 @@ static constexpr const char* allGrpProperty = "AllGroups";
 static constexpr const char* userPrivProperty = "UserPrivilege";
 static constexpr const char* userGrpProperty = "UserGroups";
 static constexpr const char* userEnabledProperty = "UserEnabled";
+static constexpr const char* userIsBootStrapProperty = "BootStrapAccount";
 
 static std::array<std::string, (PRIVILEGE_OEM + 1)> ipmiPrivIndex = {
     "priv-reserved", // PRIVILEGE_RESERVED - 0
@@ -471,6 +472,16 @@ class UserAccess
      */
     Cc removeUserFromNoneIpmiGroupUsers(const std::string& userName);
 
+    /** @brief to set user bootStrapState state
+     *
+     *  @param[in] userId - user id
+     *  @param[in] bootStrapState - bootStrapState state of the user
+     *
+     *  @return ccSuccess for success, others for failure.
+     */
+    Cc setUserBootStrapAccountState(const uint8_t userId,
+                                    const bool& bootStrapState);
+
     std::unique_ptr<boost::interprocess::named_recursive_mutex> userMutex{
         nullptr};
 
-- 
2.34.1

